### PR TITLE
MassTransit and RabbitMQ integrations were implemented. Consumer proj…

### PR DESCRIPTION
**Pull request by Ilya-Chudinovskikh**
…ect for analitics is under development.

### CODE REVIEW

**Comment by Ilya-Chudinovskikh** [comment]
Снимок экрана (25)
Снимок экрана (26)

File: WordOfTheDay.Api/Controllers/WordsController.cs [hunk 1]
```diff
@@ -60,6 +64,8 @@ public async Task<IActionResult> PostWord(Word word)
 
             await _wordsServices.PostWord(word);
 
+            await _publishEndpoint.Publish(new WordInfo(word.Id, word.Email, word.Text, word.AddTime, word.LocationLongitude, word.LocationLatitude));
```
Comment (php5art) [inline WordOfTheDay.Api/Controllers/WordsController.cs]: I would put this code to Services level, it is feel more like business logic than handling http

File: WordOfTheDay.Api/Startup.cs [hunk 2]
```diff
@@ -26,6 +25,8 @@ public void ConfigureServices(IServiceCollection services)
 
             services.AddRepositories(Configuration.GetConnectionString("WordContext"));
             services.AddDomain();
+
+            services.AddConfiguredMassTransit("amqp://guest:guest@localhost:5672");
```
Comment (php5art) [inline WordOfTheDay.Api/Startup.cs]: Put this connection to appsettings

File: WordOfTheDay.Api/Startup.cs [hunk 3]
```diff
@@ -48,6 +49,8 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
             {
                 endpoints.MapControllers();
             });
+
+
```
Comment (php5art) [inline WordOfTheDay.Api/Startup.cs]: remove redundant empty lines

File: WordOfTheDay.Repository/RepositoryServicesConfiguration.cs [hunk 4]
```diff
@@ -15,5 +15,17 @@ public static void AddRepositories(this IServiceCollection services, string conn
 
             services.AddScoped<IWordsRepository, WordsRepository>();
         }
+        public static void AddConfiguredMassTransit(this IServiceCollection services, string host)
```
Comment (php5art) [inline WordOfTheDay.Repository/RepositoryServicesConfiguration.cs]: This should not be in repository level. For me it is more for Services level

**Review by php5art — CHANGES_REQUESTED** [review]
Can you also add rabbitmq to docker-compose.yml?

File: WordOfTheDay.Api/Controllers/WordsController.cs (hunk 1 above)
Comment (Ilya-Chudinovskikh) [reply WordOfTheDay.Api/Controllers/WordsController.cs]: done

File: WordOfTheDay.Api/Startup.cs (hunk 2 above)
Comment (Ilya-Chudinovskikh) [reply WordOfTheDay.Api/Startup.cs]: done

File: WordOfTheDay.Api/Startup.cs (hunk 3 above)
Comment (Ilya-Chudinovskikh) [reply WordOfTheDay.Api/Startup.cs]: done

File: WordOfTheDay.Repository/RepositoryServicesConfiguration.cs (hunk 4 above)
Comment (Ilya-Chudinovskikh) [reply WordOfTheDay.Repository/RepositoryServicesConfiguration.cs]: done

File: WordOfTheDay.Domain/WordsServices.cs [hunk 5]
```diff
@@ -29,13 +32,15 @@ public Task<List<WordCount>> CloseWords(string email)
...
         }
-        public Task PostWord(Word word)
+        public async Task<Task> PostWord(Word word)
         {
             word.AddTime = DateTime.Today.ToUniversalTime();
 
             word.Text = word.Text.ToLower();
             word.Email = word.Email.ToLower();
 
+            await _publishEndpoint.Publish(new WordInfo(word.Id, word.Email, word.Text, word.AddTime, word.LocationLongitude, word.LocationLatitude));
```
Comment (php5art) [inline WordOfTheDay.Domain/WordsServices.cs]: Please do this after writing word to the database to be sure that word is already in database. The reason is that if repository will return an error, we woun't need to calculate any statistics using this word

File: WordOfTheDay.Repository/RepositoryServicesConfiguration.cs [hunk 6]
```diff
@@ -1,8 +1,8 @@
-﻿using Microsoft.Extensions.Configuration;
-using Microsoft.Extensions.DependencyInjection;
+﻿using Microsoft.Extensions.DependencyInjection;
 using Microsoft.EntityFrameworkCore;
 using WordOfTheDay.Repository.Entities;
 using System;
+using MassTransit;
```
Comment (php5art) [inline WordOfTheDay.Repository/RepositoryServicesConfiguration.cs]: Remove not used usages

File: WordOfTheDay.Repository/WordOfTheDay.Repository.csproj [hunk 7]
```diff
@@ -6,6 +6,10 @@
 
   <ItemGroup>
     <PackageReference Include="LinqKit" Version="1.1.27" />
+    <PackageReference Include="MassTransit" Version="7.3.1" />
```
Comment (php5art) [inline WordOfTheDay.Repository/WordOfTheDay.Repository.csproj]: Is nuget with MassTransit still needed here?

File: WordOfTheDay.Domain/WordsServices.cs [hunk 8]
```diff
@@ -29,13 +32,15 @@ public Task<List<WordCount>> CloseWords(string email)
 
             return closeWords;
         }
-        public Task PostWord(Word word)
+        public async Task<Task> PostWord(Word word)
```
Comment (php5art) [inline WordOfTheDay.Domain/WordsServices.cs]: This should return Task\<Word\>

File: WordOfTheDay.Api/Controllers/WordsController.cs [hunk 9]
```diff
@@ -2,6 +2,8 @@
 using Microsoft.AspNetCore.Mvc;
 using WordOfTheDay.Repository.Entities;
 using WordOfTheDay.Domain;
+using MassTransit;
+using SharedModelsLibrary;
```
Comment (php5art) [inline WordOfTheDay.Api/Controllers/WordsController.cs]: remove unused usages

File: WordOfTheDay.Api/WordOfTheDay.Api.csproj [hunk 10]
```diff
@@ -12,6 +12,10 @@
   </ItemGroup>
 
   <ItemGroup>
+    <PackageReference Include="MassTransit" Version="7.3.1" />
```
Comment (php5art) [inline WordOfTheDay.Api/WordOfTheDay.Api.csproj]: Do you really need this nuget in Api project?

File: WordOfTheDay.Domain/IWordsServices.cs [hunk 11]
```diff
@@ -12,7 +12,7 @@ public interface IWordsServices
 
         Task<List<WordCount>> CloseWords(string word);
 
-        Task PostWord(Word word);
+        Task<Task> PostWord(Word word);
```
Comment (php5art) [inline WordOfTheDay.Domain/IWordsServices.cs]: This method should return Task\<Word\>

File: WordOfTheDay.Repository/Entities/Word.cs [hunk 12]
```diff
@@ -1,5 +1,4 @@
-﻿using Microsoft.EntityFrameworkCore;
-using System;
+﻿using System;
```
Comment (php5art) [inline WordOfTheDay.Repository/Entities/Word.cs:1]: remove unused usages

File: WordOfTheDay.Api/appsettings.json [hunk 13]
```diff
@@ -8,6 +8,7 @@
   },
   "AllowedHosts": "*",
   "ConnectionStrings": {
-    "WordContext": "Server=(localdb)\\mssqllocaldb;Database=WordContext-03daec14-a505-48b7-8fb2-72bc85eb0fcd;Trusted_Connection=True;MultipleActiveResultSets=true"
+    "WordContext": "Server=(localdb)\\mssqllocaldb;Database=WordContext-03daec14-a505-48b7-8fb2-72bc85eb0fcd;Trusted_Connection=True;MultipleActiveResultSets=true",
+    "RabbitMQHost": "amqp://guest:guest@localhost:5672"
```
Comment (php5art) [inline WordOfTheDay.Api/appsettings.json:12]: Can you update these connections to use Docker hosted MsSql and RabbitMq?

**Review by php5art — CHANGES_REQUESTED** [review]

File: WordOfTheDay.Repository/WordOfTheDay.Repository.csproj (hunk 7 above)
Comment (Ilya-Chudinovskikh) [reply WordOfTheDay.Repository/WordOfTheDay.Repository.csproj]: done

File: WordOfTheDay.Repository/RepositoryServicesConfiguration.cs (hunk 6 above)
Comment (Ilya-Chudinovskikh) [reply WordOfTheDay.Repository/RepositoryServicesConfiguration.cs]: done

File: WordOfTheDay.Domain/WordsServices.cs (hunk 5 above)
Comment (Ilya-Chudinovskikh) [reply WordOfTheDay.Domain/WordsServices.cs]: done

File: WordOfTheDay.Domain/WordsServices.cs (hunk 8 above)
Comment (Ilya-Chudinovskikh) [reply WordOfTheDay.Domain/WordsServices.cs]: done

File: WordOfTheDay.Api/Controllers/WordsController.cs (hunk 9 above)
Comment (Ilya-Chudinovskikh) [reply WordOfTheDay.Api/Controllers/WordsController.cs]: done

File: WordOfTheDay.Api/WordOfTheDay.Api.csproj (hunk 10 above)
Comment (Ilya-Chudinovskikh) [reply WordOfTheDay.Api/WordOfTheDay.Api.csproj]: done

File: WordOfTheDay.Domain/IWordsServices.cs (hunk 11 above)
Comment (Ilya-Chudinovskikh) [reply WordOfTheDay.Domain/IWordsServices.cs]: done

File: WordOfTheDay.Repository/Entities/Word.cs (hunk 12 above)
Comment (Ilya-Chudinovskikh) [reply WordOfTheDay.Repository/Entities/Word.cs:1]: done

File: WordOfTheDay.Api/appsettings.json (hunk 13 above)
Comment (Ilya-Chudinovskikh) [reply WordOfTheDay.Api/appsettings.json:12]: Is still under development. The issue is "Microsoft.Data.SqlClient.SqlException (0x80131904): A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections." I will continue to try fixing it.

Comment (php5art) [reply WordOfTheDay.Api/appsettings.json:12]: ok, create new task for this in trello please, to track this issue

**Review by php5art — APPROVED** [review]

**Pull request merged**